pkg/authorization/admission/clusterquota/controller.go

```diff
@@ -0,0 +1,502 @@
+		deltaUsage = quota.Subtract(deltaUsage, prevUsage)
+	}
+	fmt.Printf("#### 3o\n")
+	if quota.IsZero(deltaUsage) {
+		fmt.Printf("#### 3p\n")
+		return quotas, nil
+	}
+
+	for _, index := range interestingQuotaIndexes {
+		resourceQuota := quotas[index]
+
+		hardResources := quota.ResourceNames(resourceQuota.Status.Hard)
+		requestedUsage := quota.Mask(deltaUsage, hardResources)
+		newUsage := quota.Add(resourceQuota.Status.Used, requestedUsage)
+		if allowed, exceeded := quota.LessThanOrEqual(newUsage, resourceQuota.Status.Hard); !allowed {
+			failedRequestedUsage := quota.Mask(requestedUsage, exceeded)
+			failedUsed := quota.Mask(resourceQuota.Status.Used, exceeded)
+			failedHard := quota.Mask(resourceQuota.Status.Hard, exceeded)
+			fmt.Printf("#### 3q\n")
+			return nil, admission.NewForbidden(a,
+				fmt.Errorf("Exceeded quota: %s, requested: %s, used: %s, limited: %s",
+					resourceQuota.Name,
+					prettyPrint(failedRequestedUsage),
+					prettyPrint(failedUsed),
+					prettyPrint(failedHard)))
+		}
+
+		// update to the new usage number
+		quotas[index].Status.Used = newUsage
+	}
+
+	fmt.Printf("#### 3r\n")
+	return quotas, nil
+}
+
+func (e *quotaEvaluator) evaluate(a admission.Attributes) error {
+	waiter := newAdmissionWaiter(a)
+
+	e.addWork(waiter)
+
+	// wait for completion or timeout
+	select {
+	case <-waiter.finished:
+	case <-time.After(10 * time.Second):
+		return fmt.Errorf("timeout")
+	}
+
+	return waiter.result
+}
+
+func (e *quotaEvaluator) addWork(a *admissionWaiter) {
+	e.workLock.Lock()
+	defer e.workLock.Unlock()
+
+	ns := a.attributes.GetNamespace()
+	// this Add can trigger a Get BEFORE the work is added to a list, but this is ok because the getWork routine
+	// waits the worklock before retrieving the work to do, so the writes in this method will be observed
+	e.queue.Add(ns)
+
+	if e.inProgress.Has(ns) {
+		e.dirtyWork[ns] = append(e.dirtyWork[ns], a)
+		return
+	}
+
+	e.work[ns] = append(e.work[ns], a)
+}
+
+func (e *quotaEvaluator) completeWork(ns string) {
+	e.workLock.Lock()
+	defer e.workLock.Unlock()
+
+	e.queue.Done(ns)
+	e.work[ns] = e.dirtyWork[ns]
+	delete(e.dirtyWork, ns)
+	e.inProgress.Delete(ns)
+}
+
+func (e *quotaEvaluator) getWork() (string, []*admissionWaiter) {
+	uncastNS, _ := e.queue.Get()
+	ns := uncastNS.(string)
+
+	e.workLock.Lock()
+	defer e.workLock.Unlock()
+	// at this point, we know we have a coherent view of e.work.  It is entirely possible
+	// that our workqueue has another item requeued to it, but we'll pick it up early.  This ok
+	// because the next time will go into our dirty list
+
+	work := e.work[ns]
+	delete(e.work, ns)
+	delete(e.dirtyWork, ns)
+
+	if len(work) != 0 {
+		e.inProgress.Insert(ns)
+		return ns, work
+	}
+
+	e.queue.Done(ns)
+	e.inProgress.Delete(ns)
+	return ns, []*admissionWaiter{}
+}
+
+func (e *quotaEvaluator) getQuotas(namespace string) ([]api.ResourceQuota, error) {
+	// trust me to do somethign clever here with a pre-computed cache ala project cache
```

Comment:
i imagine this cache is built against shared informers, so its probably no worse.